// qenya/middlewares/render.js
'use strict';exports.__esModule = true;var _regenerator = require('babel-runtime/regenerator');var _regenerator2 = _interopRequireDefault(_regenerator);var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}exports.default = function () {var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(ctx, next) {return _regenerator2.default.wrap(function _callee$(_context) {while (1) {switch (_context.prev = _context.next) {case 0:
            ctx.Render = {
              success: function success(data, msg) {
                ctx.body = {
                  data: data };

                if (msg) {
                  ctx.body.msg = msg;
                }
              },

              fail: function fail(err) {
                ctx.body = {
                  error: err };

              },

              notFound: function notFound() {
                ctx.body = {
                  error: '请求的资源不存在' };

              },

              noAuth: function noAuth() {
                ctx.body = {
                  error: '没有权限' };

                ctx.response.status = 401;
              },

              expired: function expired() {
                ctx.body = {
                  error: '登录已过期' };

                ctx.response.status = 401;
              } };_context.next = 3;return (


              next());case 3:case 'end':return _context.stop();}}}, _callee, this);}));return function (_x, _x2) {return _ref.apply(this, arguments);};}();
